### PR TITLE
Simd v2

### PR DESCRIPTION
**Pull request by CarlVerret**
This PR introduce SIMD operation for UTF-16 string parsing.  It improves performance a lot.  From now on performance of UTF-16 parsing is equivalent (if not faster) as UTF-8.  

In some cases, it is up to 8 to 9 times faster than standard library.

|                              Method |           FileName |      Mean |     Error |    StdDev |       Min | Ratio | MFloat/s |     MB/s |
|------------------------------------ |------------------- |----------:|----------:|----------:|----------:|------:|---------:|---------:|
| 'FastFloat.TryParseDouble() - UTF8' |    data/canada.txt |  4.690 ms | 0.0053 ms | 0.0050 ms |  4.683 ms |  0.12 |    23.73 |   445.85 |
|          FastFloat.TryParseDouble() |    data/canada.txt |  4.586 ms | 0.0167 ms | 0.0148 ms |  4.559 ms |  0.12 |    24.38 |   458.00 |
|          'ParseNumberString() only' |    data/canada.txt |  2.472 ms | 0.0395 ms | 0.0370 ms |  2.411 ms |  0.07 |    46.10 |   866.13 |
|                      Double.Parse() |    data/canada.txt | 37.537 ms | 0.2360 ms | 0.2208 ms | 37.159 ms |  1.00 |     2.99 |    56.19 |
|                                     |                    |           |           |           |           |       |          |          |
| 'FastFloat.TryParseDouble() - UTF8' |      data/mesh.txt |  1.833 ms | 0.0004 ms | 0.0003 ms |  1.832 ms |  0.27 |    39.86 |   338.41 |
|          FastFloat.TryParseDouble() |      data/mesh.txt |  1.834 ms | 0.0003 ms | 0.0003 ms |  1.834 ms |  0.27 |    39.81 |   338.05 |
|          'ParseNumberString() only' |      data/mesh.txt |  1.168 ms | 0.0021 ms | 0.0020 ms |  1.164 ms |  0.17 |    62.71 |   532.43 |
|                      Double.Parse() |      data/mesh.txt |  6.850 ms | 0.0375 ms | 0.0351 ms |  6.788 ms |  1.00 |    10.76 |    91.34 |
|                                     |                    |           |           |           |           |       |          |          |
| 'FastFloat.TryParseDouble() - UTF8' | data/synthetic.txt |  5.295 ms | 0.0131 ms | 0.0109 ms |  5.283 ms |  0.11 |    28.39 |   534.13 |
|          FastFloat.TryParseDouble() | data/synthetic.txt |  5.147 ms | 0.0154 ms | 0.0144 ms |  5.118 ms |  0.11 |    29.31 |   551.44 |
|          'ParseNumberString() only' | data/synthetic.txt |  2.655 ms | 0.0025 ms | 0.0019 ms |  2.653 ms |  0.05 |    56.54 |  1063.78 |
|                      Double.Parse() | data/synthetic.txt | 48.744 ms | 0.3320 ms | 0.3105 ms | 48.283 ms |  1.00 |     3.11 |    58.45 |

### CODE REVIEW

File: csFastFloat/FastDoubleParser.cs [hunk 1]
```diff
@@ -445,13 +445,10 @@ public static unsafe double ParseDouble(byte* first, byte* last, out int charact
     /// <param name="characters_consumed">number of characters consumed while parsing</param>
     /// <param name="expectedFormat">allowed styles for the input string</param>
     /// <param name="decimal_separator">decimal separator to be used</param>
+    /// <param name="value">out : reference to double variable to hold parsed value</param>
```
Comment (lemire) [inline csFastFloat/FastDoubleParser.cs:448]: could be spelled out:

"to hold parsed value" -> "to hold the parsed value"

File: csFastFloat/Utils/Utils.cs [hunk 2]
```diff
@@ -241,13 +249,58 @@ static int Log2SoftwareFallback(uint value)
...
 #else
-      return *((float*)&value);
+      => *((float*)&value);
 #endif
+
+
+
+
+
+#if HAS_INTRINSICS
```
[the remark this answers is not in the document]
Comment (lemire) [reply csFastFloat/Utils/Utils.cs:262]: I'd expect a check like this one: 

unnamed-3

[the remark this answers is not in the document]
Comment (lemire) [reply csFastFloat/Utils/Utils.cs:262]: Specifically, I'd expect something like using System.Runtime.Intrinsics.X86.Sse42.IsSupported or something of the sort.

[the remark this answers is not in the document]
Comment (lemire) [reply csFastFloat/Utils/Utils.cs:262]: E.g., see this code: https://github.com/RossNordby/scratchpad/blob/fe8aeea5232b49632c040730ee200ba6a0994113/IntrinsicsTesting/IntrinsicsTesting/Bundles/Wide.cs

[the remark this answers is not in the document]
Comment (CarlVerret) [reply csFastFloat/Utils/Utils.cs:262]: Actually, I feared that Sse42.IsSupported won't compile in .net standard.  That why I used HAS_INTRINSICS.  I fully agree to use something better if available.

[the remark this answers is not in the document]
Comment (lemire) [reply csFastFloat/Utils/Utils.cs:262]: So I agree that the code I linked to is not great. But I was just trying to illustrate the point.

[the remark this answers is not in the document]
Comment (lemire) [reply csFastFloat/Utils/Utils.cs:262]: Let me be clearer. You need (1) to check that intrinsics are supported (2) and if so, and only if so, then check that sse42 is supported.

File: csFastFloat/Utils/Utils.cs [hunk 3]
```diff
@@ -241,13 +249,58 @@ static int Log2SoftwareFallback(uint value)
...
+      {
+        return false;
+      }
+
+      // @Credit  AQRIT
+      // https://stackoverflow.com/questions/66371621/hardware-simd-parsing-in-c-sharp-performance-improvement/66430672
+      Vector128<byte> mul1 = Vector128.Create(0x14C814C8, 0x010A0A64, 0, 0).AsByte();
+      Vector128<short> mul2 = Vector128.Create(0x00FA61A8, 0x0001000A, 0, 0).AsInt16();
+
+      Vector128<byte> vb = Sse2.PackUnsignedSaturate(raw, raw);
```
[the remark this answers is not in the document]
Comment (CarlVerret) [reply csFastFloat/Utils/Utils.cs:291]: forgive me if I'm wrong, but to use a again I must substract short.minvalue from it : 
  a = Sse41.Subtract(a, Vector128.Create((short)(short.MinValue))); 
unless I'm stuck with "-32678 biased" values in my 'a' vector.  Maybe there's a trick I don't know about.

File: csFastFloat/Utils/Utils.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (CarlVerret) [reply csFastFloat/Utils/Utils.cs:262]: I added a guard for both Code and unit tests.

File: csFastFloat/FastDoubleParser.cs (hunk 1 above)
Comment (CarlVerret) [reply csFastFloat/FastDoubleParser.cs]: done

File: Benchmark/FastParserBenchmark.cs [hunk 4]
```diff
@@ -12,157 +12,139 @@
...
 
   //[MemoryDiagnoser]
   [SimpleJob(RuntimeMoniker.NetCoreApp50)]
-  [Config(typeof(Config))]
-  public class FFBencmark
-  {
-    private string[] _lines;
-    private byte[][] _linesUtf8;
+[Config(typeof(Config))]
+public class FFBencmark
```
[the remark this answers is not in the document]
Comment (CarlVerret) [reply Benchmark/FastParserBenchmark.cs]: oups!   done

File: csFastFloat/Utils/Utils.cs [hunk 5]
```diff
@@ -241,13 +249,66 @@ static int Log2SoftwareFallback(uint value)
...
+      {
+        value = 0;
+        return false;
+      }
+
+
+      value = 0;
+      Vector128<short> raw = Sse41.LoadDquVector128((short*)start);
+      Vector128<short> ascii0 = Vector128.Create((short)(48 + short.MinValue));
+      Vector128<short> after_ascii9 = Vector128.Create((short)(short.MinValue + 9));
```
[the remark this answers is not in the document]
Comment (lemire) [reply csFastFloat/Utils/Utils.cs:285]: I would check whether they are hoisted by looking at the assembly. I think @CarlVerret did this check.

File: csFastFloat/Utils/Utils.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lemire) [reply csFastFloat/Utils/Utils.cs:299]: @CarlVerret So -32678 is 0b1000000000000000 in binary. So -32678+1 is 0b1000000000000001 in binary. And so forth. So you just need to extract the low bytes of each 16-bit word. E.g., the following will give you 1,1,1,1...

```C++
__m128i x =  _mm_set1_epi16(-32768+1);
x=_mm_shuffle_epi8(x,_mm_setr_epi8(0,2,4,6,8,10,12,14,0,2,4,6,8,10,12,14));
```

You should be able to do the same with `Ssse3.Shuffle` and then use the usual create:

https://docs.microsoft.com/en-us/dotnet/api/system.runtime.intrinsics.vector128.create?view=net-5.0#System_Runtime_Intrinsics_Vector128_Create_System_SByte_System_SByte_System_SByte_System_SByte_System_SByte_System_SByte_System_SByte_System_SByte_System_SByte_System_SByte_System_SByte_System_SByte_System_SByte_System_SByte_System_SByte_System_SByte_

File: csFastFloat/Utils/Utils.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lemire) [reply csFastFloat/Utils/Utils.cs:285]: > I guess the JIT isn't able to hoist these "constants" outside the loop.

I think that's an incorrection assumption, but it is quite true that one should check. I think this check was done.

**Pull request merged**